### PR TITLE
Specialize storage_offset - Does not cover automatic dynamic

### PR DESCRIPTION
**Pull request by voznesenskym**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #104204

cc @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @chenyang78

### CODE REVIEW

File: torch/_dynamo/variables/tensor.py [hunk 1]
```diff
@@ -227,6 +231,10 @@ def var_getattr(self, tx, name):
             result = ConstantVariable(self.is_quantized, **options)
         elif name == "is_sparse" and self.is_sparse is not None:
             result = ConstantVariable(self.is_sparse, **options)
+        elif name == "storage_offset":
+            return variables.LambdaVariable(
+                lambda *args, **kwargs: ConstantVariable(self.storage_offset)
```
Comment (wconstab) [inline torch/_dynamo/variables/tensor.py:236]: is **options missing here relative to the call_method impl returning `ConstantVariable(self.storage_offset, **options)` below?

Comment (voznesenskym) [reply torch/_dynamo/variables/tensor.py:236]: LambdaVariable covers this!

LambdaVariable here gets a  `.add_options(self)`, And its call_function passes options down:

```
    def call_function(
        self, tx, args: "List[VariableTracker]", kwargs: "Dict[str, VariableTracker]"
    ) -> "VariableTracker":
        return self.fn(*args, **kwargs).add_options(self)
```

File: torch/_dynamo/variables/tensor.py [hunk 2]
```diff
@@ -150,6 +152,8 @@ def specialize(value: torch.Tensor):
                     if value.is_contiguous(memory_format=x)
                 ]
             )
+            props["storage_offset"] = value.storage_offset()
```
Comment (wconstab) [inline torch/_dynamo/variables/tensor.py:155]: confirmed: free_symbols already checks storage_offset so this looks ok

File: torch/csrc/dynamo/guards.cpp [hunk 3]
```diff
@@ -128,6 +133,11 @@ class TensorCheck {
         return fail_reason.str();
       }
     }
+    if (offset_.has_value() && offset_.value() != v.storage_offset()) {
+      fail_reason << "storage_offset mismatch, expected " << v.storage_offset()
```
Comment (wconstab) [inline torch/csrc/dynamo/guards.cpp:137]: expected and actual are reversed?

File: torch/csrc/dynamo/guards.cpp [hunk 4]
```diff
@@ -141,9 +151,11 @@ class TensorCheck {
   // necessarily capture device indices correctly.
   at::DeviceIndex device_index_;
   bool requires_grad_;
-  // NB: These are unset if dynamic shapes is enabled.
+  // NB: These are always set, but will contains empty optionals if dynamic
```
Comment (wconstab) [inline torch/csrc/dynamo/guards.cpp:154]: typo contain (no 's')

File: torch/csrc/dynamo/guards.cpp (hunk 3 above)
Comment (voznesenskym) [reply torch/csrc/dynamo/guards.cpp:137]: yea

File: torch/_dynamo/variables/tensor.py (hunk 2 above)
Comment (voznesenskym) [reply torch/_dynamo/variables/tensor.py:155]: Yea.

File: torch/csrc/dynamo/guards.cpp (hunk 3 above)
Comment (voznesenskym) [reply torch/csrc/dynamo/guards.cpp:137]: Good catch

Comment (voznesenskym) [reply torch/csrc/dynamo/guards.cpp:137]: Oh I think it was wrong for sizes and strides too.

**Review by wconstab — APPROVED** [review]
lgtm, probably worth waiting for ci/trunk on this one just to be safe

File: torch/csrc/dynamo/guards.cpp [hunk 5]
```diff
@@ -119,15 +124,20 @@ class TensorCheck {
...
         fail_reason << "size mismatch at index " << i << ". expected "
-                    << known_size.value() << ", actual " << sizes[i];
+                    << sizes[i] << ", actual " << known_size.value();
         return fail_reason.str();
       }
       if (known_stride.has_value() && known_stride.value() != strides[i]) {
         fail_reason << "stride mismatch at index " << i << ". expected "
-                    << known_stride.value() << ", actual " << strides[i];
+                    << strides[i] << ", actual " << known_stride.value();
         return fail_reason.str();
```
Comment (voznesenskym) [inline torch/csrc/dynamo/guards.cpp:133]: wait, did I confuse myself here? hmmm

**Comment by jeanschmidt** [comment]
Hi, 

it seems this PR broke one of the CI jobs [https://github.com/pytorch/pytorch/actions/runs/5386259326/jobs/9776853834#step:15:3546]("[cuda11.8-py3.10-gcc7-sm86 / test (inductor_torchbench, 1, 1, linux.g5.4xlarge.nvidia.gpu)](https://github.com/pytorch/pytorch/actions/runs/5386259326/jobs/9776853834#logs)") 

Can we easily FF this?

**Comment by voznesenskym** [comment]
> Hi @voznesenskym,
> 
> it seems this PR broke one of the CI jobs [https://github.com/pytorch/pytorch/actions/runs/5386259326/jobs/9776853834#step:15:3546]("[cuda11.8-py3.10-gcc7-sm86 / test (inductor_torchbench, 1, 1, linux.g5.4xlarge.nvidia.gpu)](https://github.com/pytorch/pytorch/actions/runs/5386259326/jobs/9776853834#logs)")
> 
> Can we easily FF this?

vision mask rcnn is usually unstable. I don't trust this CI signal, because it did not come up on the PR.

**Comment by ezyang** [comment]
But it did show up on the PR...

image
